proc_macros/src/grouped_ordering.rs

```diff
@@ -0,0 +1,186 @@
+
+    quote! {
+        impl #crate_name::GroupedOrdering for #name {
+            type Group = #group_enum_name;
+
+            fn compare(
+                &self,
+                a: &#group_enum_name,
+                b: &#group_enum_name,
+            ) -> Ordering {
+                self.index_lookup[a].cmp(&self.index_lookup[b])
+            }
+        }
+    }
+}
+
+fn get_impl_try_from(grouped_ordering_spec: &GroupedOrderingSpec, group_enum_name: &Ident) -> proc_macro2::TokenStream {
+    let name = &grouped_ordering_spec.name;
+    let num_groups = grouped_ordering_spec.groups.len();
+
+    quote! {
+        impl TryFrom<[#group_enum_name; #num_groups]> for #name {
+            type Error = String;
+
+            fn try_from(groups: [#group_enum_name; #num_groups]) -> Result<Self, Self::Error> {
+                let index_lookup: std::collections::HashMap<#group_enum_name, usize> = groups
+                    .into_iter()
+                    .enumerate()
+                    .map(|(index, group)| (group, index))
+                    .collect();
+                if index_lookup.len() < #num_groups {
+                    return Err("Found duplicate groups".to_owned());
+                }
+                Ok(Self {
+                    groups,
+                    index_lookup,
+                })
+            }
+        }
+    }
+}
+
+fn get_impl_default(grouped_ordering_spec: &GroupedOrderingSpec, group_enum_name: &Ident) -> proc_macro2::TokenStream {
+    let name = &grouped_ordering_spec.name;
+    let qualified_groups = grouped_ordering_spec.groups.iter().map(|group| {
+        quote! {
+            #group_enum_name::#group
+        }
+    }).collect::<Vec<_>>();
+
+    quote! {
+        impl Default for #name {
+            fn default() -> Self {
+                Self::try_from([#(#qualified_groups),*]).unwrap()
+            }
+        }
+    }
+}
+
+fn get_impl_deserialize(grouped_ordering_spec: &GroupedOrderingSpec, group_enum_name: &Ident) -> proc_macro2::TokenStream {
+    let name = &grouped_ordering_spec.name;
+    let num_groups = grouped_ordering_spec.groups.len();
+
+    quote! {
+        impl<'de> serde::Deserialize<'de> for #name {
+            fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+            where
+                D: serde::Deserializer<'de>,
+            {
+                let array = <[#group_enum_name; #num_groups] as serde::Deserialize>::deserialize(deserializer)?;
+                use serde::de::Error;
+                Self::try_from(array).map_err(|_| D::Error::custom("Expected all variants"))
+            }
+        }
+    }
+}
+
+fn get_instantiate_macro_definition(grouped_ordering_spec: &GroupedOrderingSpec, group_enum_name: &Ident) -> proc_macro2::TokenStream {
+    let macro_name = format_ident!("{}", grouped_ordering_spec.name.to_string().to_snake_case());
+    let name = &grouped_ordering_spec.name;
+
+    quote! {
+        #[allow(unused_macros)]
```

Comment:
Macro scoping is confusing, I tried mimicking the tree-sitter-grep fully-populated collection macros and adding `#[macro_export]` to this but then my tests complained that multiple macros of the same name were being attempted to be added to the root scope or whatever (so correspondingly it's probably wrong that I'm using `#[macro_export]` in tree-sitter-grep) so removed that. Then it was warning about unused macros when I didn't actually make use of this generated macro so added this `#[allow(unused_macros)]`. Not clear on whether one would run into trickiness/difficulty if you wanted to eg define a grouped-ordering type in its own file and then export the generated instantiation macro?